trin-execution/src/subcommands/era2.rs

```diff
@@ -0,0 +1,243 @@
+use std::sync::Arc;
+
+use alloy_consensus::EMPTY_ROOT_HASH;
+use alloy_rlp::Decodable;
+use anyhow::{ensure, Error};
+use e2store::era2::{
+    AccountEntry, AccountOrStorageEntry, Era2, StorageEntry, StorageItem, MAX_STORAGE_ITEMS,
+};
+use eth_trie::{EthTrie, Trie};
+use ethportal_api::Header;
+use parking_lot::Mutex;
+use revm_primitives::{keccak256, B256, KECCAK_EMPTY, U256};
+use tracing::info;
+
+use crate::{
+    cli::{ExportStateConfig, ImportStateConfig},
+    era::manager::EraManager,
+    evm::block_executor::BLOCKHASH_SERVE_WINDOW,
+    execution::TrinExecution,
+    storage::{account::Account, account_db::AccountDB},
+};
+
+pub struct StateExporter {
+    pub trin_execution: TrinExecution,
+    exporter_config: ExportStateConfig,
+}
+
+impl StateExporter {
+    pub fn new(trin_execution: TrinExecution, exporter_config: ExportStateConfig) -> Self {
+        Self {
+            trin_execution,
+            exporter_config,
+        }
+    }
+
+    pub fn export_state(&mut self, header: Header) -> anyhow::Result<()> {
+        ensure!(
+            header.state_root == self.trin_execution.get_root()?,
+            "State root mismatch fro block header we are trying to export"
+        );
+        info!(
+            "Exporting state from block number: {} with state root: {}",
+            header.number, header.state_root
+        );
+        let mut era2 = Era2::create(self.exporter_config.path_to_era2.clone(), header)?;
+        info!("Era2 initiated");
+        info!("Trie leaf iterator initiated");
+        let mut accounts_processed = 0;
+        for key_hash_and_leaf_value in self.trin_execution.database.trie.lock().iter() {
+            let (raw_account_hash, account_state) = key_hash_and_leaf_value?;
+            let account_hash = B256::from_slice(&raw_account_hash);
+
+            let account_state: Account = Decodable::decode(&mut account_state.as_slice())?;
+            let bytecode = if account_state.code_hash != KECCAK_EMPTY {
+                self.trin_execution
+                    .database
+                    .db
+                    .get(account_state.code_hash)?
+                    .expect("If code hash is not empty, code must be present")
+            } else {
+                vec![]
+            };
+
+            let mut storage: Vec<StorageItem> = vec![];
```

Comment:
nit: I don't know if this is big problem, but I don't like that we are storing all storage items in memory, instead of exporting them to file. Don't know how big storage can get and if this is an issue (at least on weaker machines).

I know that this is needed in order to write `storage_count` in the `AccountEntry` first. The possible solutions would be:

1. iterate entire trie once and count number of items, write `AccountEntry`, iterate second time and write as you go
    - considering that reading from db is most likely the slowest part, this would make export twice as slow
2. Start processing as we do now.
    1. if we finish processing entire storage trie before getting to `MAX_STORAGE_ITEMS`, do as we do now (in which case, there will be only one StorageEntry)
    2. If we process `MAX_STORAGE_ITEMS`, stop, and do some variation of 1. from above.
    - this approach gives us best performance in majority of cases (single StorageEntry), but still handles big storage tries. but it's not elegant, and questionable how much it would actually slow things down

Overall, I think current implementation is good for now, and if we notice issue with later (bigger) blocks, we can evaluate if improvement is needed. So no action is expected here (except if you want to create an issue to document this and provide your opinion).

Reply:
If slower machines can't even run an execution client, I am not sure why they should be able to generate a `.era2` file.

They should minimum have 16GB of ram, and if that is the case there should be no problem, and if they need more they can add a swap file.

This isn't an issue, creating `.era2` files is already
- expensive
- slow
- niche

99% of people won't be generating these files, they will be using the files we generated.

So I don't think we should optimize this for the lowest case, they can use a .era2 file we generated or use a functional machine.